rust/kernel/src/sync.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::cell::UnsafeCell;
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+
+use crate::bindings;
+
+pub struct Mutex<T: ?Sized> {
+    lock: bindings::mutex,
```

Comment:
`struct mutex` contains a `struct list_head`, which is a self-referential data structure. So we need to pin Mutex before we can safely initialise and use it.